apollo-federation/src/operation/selection_map.rs

```diff
@@ -0,0 +1,647 @@
+    pub(crate) fn first(&self) -> Option<&Selection> {
+        self.selections.first()
+    }
+
+    /// Computes the hash of a selection key.
+    fn hash(&self, key: SelectionKey<'_>) -> u64 {
+        self.hash_builder.hash_one(key)
+    }
+
+    /// Returns true if the given key exists in the map.
+    pub(crate) fn contains_key(&self, key: SelectionKey<'_>) -> bool {
+        let hash = self.hash(key);
+        self.table
+            .find(hash, key_eq(&self.selections, key))
+            .is_some()
+    }
+
+    /// Returns true if the given key exists in the map.
+    pub(crate) fn get(&self, key: SelectionKey<'_>) -> Option<&Selection> {
+        let hash = self.hash(key);
+        let bucket = self.table.find(hash, key_eq(&self.selections, key))?;
+        Some(&self.selections[bucket.index])
+    }
+
+    pub(crate) fn get_mut(&mut self, key: SelectionKey<'_>) -> Option<SelectionValue<'_>> {
+        let hash = self.hash(key);
+        let bucket = self.table.find_mut(hash, key_eq(&self.selections, key))?;
+        Some(SelectionValue::new(&mut self.selections[bucket.index]))
+    }
+
+    /// Insert a selection into the map.
+    fn raw_insert(&mut self, hash: u64, value: Selection) -> &mut Selection {
+        let index = self.selections.len();
+
+        self.table
+            .insert_unique(hash, Bucket { index, hash }, |existing| existing.hash);
+
+        self.selections.push(value);
+        &mut self.selections[index]
+    }
+
+    /// Resets and rebuilds the hash table.
+    ///
+    /// Preconditions:
+    /// - The table must have enough capacity for `self.selections.len()` elements.
+    fn rebuild_table_no_grow(&mut self) {
+        assert!(self.table.capacity() >= self.selections.len());
+        self.table.clear();
+        for (index, selection) in self.selections.iter().enumerate() {
+            let hash = self.hash(selection.key());
+            self.table
+                .insert_unique(hash, Bucket { index, hash }, |existing| existing.hash);
+        }
+    }
+
+    /// Decrements all the indices in the table starting at `pivot`.
+    fn decrement_table(&mut self, pivot: usize) {
+        for bucket in self.table.iter_mut() {
+            if bucket.index >= pivot {
+                bucket.index -= 1;
+            }
+        }
+    }
+
+    pub(crate) fn insert(&mut self, value: Selection) {
+        let hash = self.hash(value.key());
+        self.raw_insert(hash, value);
+    }
+
+    /// Remove a selection from the map. Returns the selection and its numeric index.
+    pub(crate) fn remove(&mut self, key: SelectionKey<'_>) -> Option<(usize, Selection)> {
+        let hash = self.hash(key);
+        let entry = self
+            .table
+            .find_entry(hash, key_eq(&self.selections, key))
+            .ok()?;
+        let (bucket, _) = entry.remove();
+        let selection = self.selections.remove(bucket.index);
+        self.decrement_table(bucket.index);
```

Comment:
I think we can be slight more efficient here. Rather than iterating through the entire table, iterate through `self.selections` starting at `bucket.index`, get that key's bucket, and decrement. Roughly this:
```rust
let selection = self.selections.remove(bucket.index);
for selection in &self.selection[bucket.index..] {
  self.table.find_mut(hash, key_eq(&self.selections, key)).unwrap().index -= 1;
}
```

Reply:
It's a good thought, but I think it's not a slam-dunk improvement... The table is an array internally, so going through it linearly from beginning to end is going to be very fast, while starting from the selections will require jumping around in the memory. I think it's sometimes going to be better and sometimes worse. I like the simplicity of iterating the whole thing.